src/tyre.mli

```diff
@@ -29,28 +29,51 @@ val dim : (int * int) Tyre.t
 
 *)
 
-type 'a t
+type non_evaluable
+type evaluable
+
+type ('evaluable, 'a) t
 (** A typed regular expression.
 
-    The type variable is the type of the returned value when the typed regular expression (tyregex) is executed.
+    The type variable is the type of the returned value when the typed regular
+    expression (tyregex) is executed.
+
+    For example [tyre : (_, int) t] can be used to return an [int]. In the rest
+    of the documentation, we will use «[tyre]» to designate a value of type
+    {!t}.
+
+    A value of type [(evaluable, a) t] can be used with the {!eval} function to
+    returns a string [s] such that [exec (compile tyre) s = v]. We call such a
+    value an {!expression}.
 
-    For example [tyre : int t] can be used to return an [int]. In the rest of the documentation, we will use «[tyre]» to designate a value of type {!t}.
+    A value of type [(non_evaluable, a) t] can only be used with functions that
+    match a string, it can't be used to produce an example string. Because its
+    only usable for matching, it is called a {!pattern}.
 *)
 
+type 'a pattern = (non_evaluable, 'a) t
+(** A regexp only usable for matching *)
+
+type 'a expression = (evaluable, 'a) t
+(** A regexp usable both for matching and evaluating. *)
+
+val unlift: 'a expression -> 'a pattern
+(** Turn an expression into a pattern. *)
```

Comment:
Have you observed code where this is needed ? Given the API, it seems like this would never be used.